### ISSUE TITLE
LayoutLMv2/XLM missing implementation for Document Classification

### ISSUE THREAD

**Issue by dcyoung** [microsoft/unilm#329]
Looking through the [recently posted codebase](https://github.com/microsoft/unilm/tree/master/layoutlmft) I noted that [modeling_layoutlmv2.py](https://github.com/microsoft/unilm/blob/master/layoutlmft/layoutlmft/models/layoutlmv2/modeling_layoutlmv2.py) does not contain an implementation for document or sequence classification. The original LayoutLMv2 paper describes: 

```
Fine-tuning LayoutLMv2 for Document Image Classification This task depends on high-level
visual information, thereby we leverage the image features explicitly in the fine-tuning. We pool the
visual embeddings into a global pre-encoder feature, and pool the visual part of LayoutLMv2 output
representations into a global post-encoder feature. The pre and post-encoder features along with the
[CLS] output feature are concatenated and fed into the final classification layer
```
Are there any examples of fine-tuning LayoutLMv2 or LayoutXLM for document classification? If not, do you plan to add any such examples in the future?

If you do not intend to add such an example, can you provide some insight for those looking to add this functionality? Would the implementation follow the pattern present in the [deprecated legacy code for LayoutLM](https://github.com/microsoft/unilm/blob/master/layoutlm/deprecated/layoutlm/modeling/layoutlm.py#L243).

**Comment by wolfshow** [comment]
@dcyoung  Yes, you may follow the LayoutLMv1 implementation to fine-tune the [CLS] output.

**Comment by dcyoung** [comment]
> @dcyoung Yes, you may follow the LayoutLMv1 implementation to fine-tune the [CLS] output.

Thanks for the response @wolfshow  

I have a basic version working well using only the existing LayoutLMv2Model's pooler outputs (`[CLS]`). I'm now attempting to add the global pre and post encoder visual features described in the paper (quoted here in the original post). I am not sure how to handle the described pooling. 

I modified the existing forward implementation of `LayoutLMv2Model` to return the visual embeddings like so:
```python
class LayoutLMv2Model(LayoutLMv2PreTrainedModel):
      ...
      def forward(..):
         ....
         return BaseModelOutputWithPoolingAndCrossAttentions(...), visual_emb
```
Then my derived model attempts to leverage these embeddings:
```python
class LayoutLMv2ForSequenceClassification(LayoutLMv2PreTrainedModel):
  ...
  def forward(...):
     ...
     outputs, visual_emb = self.layoutlmv2(...)
     # pool the visual embeddings into a global pre-encoder feature
     pre_encoder_visual_feature = MyCustomPooler(visual_emb)  # Need help with implementing this pooler
     # pool the visual part of LayoutLMv2 output representations into a global post-encoder feature
     visual_output_representation = outputs.last_hidden_state[:, -49:]
     post_encoder_visual_output_feature = MyCustomPooler(visual_output_representation) # Need direction on implementing this pooler
     # Concatenate the pre and post-encoder features along with the [CLS] output feature
     cls_output_feature = outputs.pooler_output
     combined_pooled_features =  torch.cat([pre_encoder_visual_feature,cls_output_feature,post_encoder_visual_output_feature,],dim=1,)
     logits = self.classifier(combined_pooled_features)
     ....
```

The `[CLS]` output feature has dims `(batch_size, 768)`, while the visual embeddings and outputs have dimension `(batch_size, 49, 768)`. I'm not sure how to perform the pooling for these visual features. 

- a 2D spatial pooling doesn't seem appropriate, given the outputs do not represent image data directly
- naively flattening the 2D features would yield a vector of length 49x768 =37,632. I would think such a large vector on either side of the [CLS] feature in the concatenated output would drown out the [CLS] feature. 
- I'm lacking intuition on how to reduce the dimensionality of the 2D features

Can you provide direction and possibly some intuition about how to implement the `MyCustomPooler` noted above?

**Comment by dhananjayd232** [comment]
> > @dcyoung Yes, you may follow the LayoutLMv1 implementation to fine-tune the [CLS] output.
> 
> Thanks for the response @wolfshow
> 
> I have a basic version working well using only the existing LayoutLMv2Model's pooler outputs (`[CLS]`). I'm now attempting to add the global pre and post encoder visual features described in the paper (quoted here in the original post). I am not sure how to handle the described pooling.
> 
> I modified the existing forward implementation of `LayoutLMv2Model` to return the visual embeddings like so:
> 
> ```python
> class LayoutLMv2Model(LayoutLMv2PreTrainedModel):
>       ...
>       def forward(..):
>          ....
>          return BaseModelOutputWithPoolingAndCrossAttentions(...), visual_emb
> ```
> 
> Then my derived model attempts to leverage these embeddings:
> 
> ```python
> # Need help with implementing this pooler
> class VisualPooler(nn.Module):
>     def __init__(self):
>         super().__init__()
>         self.dense = nn.Linear(49, 1)
>         self.activation = nn.Tanh()
> 
>     def forward(self, input):
>         pooled_output = self.dense(input.permute(0, 2, 1)).permute(0, 2, 1)
>         pooled_output = torch.squeeze(pooled_output, 1)
>         pooled_output = self.activation(pooled_output)
>         return pooled_output
> 
> class LayoutLMv2ForSequenceClassification(LayoutLMv2PreTrainedModel):
>     def __init__(self, config):
>         super().__init__(config)
>         self.num_labels = config.num_labels
>         self.layoutlmv2 = LayoutLMv2Model(config)
>         self.pre_encoder_visual_emb_pooler = VisualPooler()
>         self.visual_output_pooler = VisualPooler()
>         self.dropout = nn.Dropout(config.hidden_dropout_prob)
>         self.classifier = nn.Linear(3 * config.hidden_size, config.num_labels)
>         self.init_weights()
>   ...
>   def forward(...):
>      ...
>      outputs, visual_emb = self.layoutlmv2(...)
>      # pool the visual embeddings into a global pre-encoder feature
>      pre_encoder_visual_feature = self.pre_encoder_visual_emb_pooler(visual_emb)
>      # pool the visual part of LayoutLMv2 output representations into a global post-encoder feature
>      visual_output_representation = outputs.last_hidden_state[:, -49:]
>      post_encoder_visual_output_feature = self.visual_output_pooler(visual_output_representation)
>      # Concatenate the pre and post-encoder features along with the [CLS] output feature
>      cls_output_feature = outputs.pooler_output
>      combined_pooled_features =  torch.cat([pre_encoder_visual_feature,cls_output_feature,post_encoder_visual_output_feature,],dim=1,)
>      # Classify
>      classifier_input = self.dropout(combined_pooled_features)
>      logits = self.classifier(classifier_input)
>      ....
> ```
> 
> The `[CLS]` output feature has dims `(batch_size, 768)`, while the visual embeddings and outputs have dimension `(batch_size, 49, 768)`. I'm not sure how to perform the pooling for these visual features.
> 
> * a 2D spatial pooling doesn't seem appropriate, given the outputs do not represent image data directly
> * naively flattening the 2D features would yield a vector of length 49x768 =37,632. I would think such a large vector on either side of the [CLS] feature in the concatenated output would drown out the [CLS] feature.
> * I'm lacking intuition on how to reduce the dimensionality of the 2D features
> 
> Can you provide direction and possibly some intuition about how to better implement the `VisualPooler` noted above?

Hi @dcyoung,
Thanks for posting this issue, I also want to use LayoutLMv2 for document classification.
Can you please share your researched code snippet for my reference.

Thanks,
Dhananjay
